cidc_schemas/template.py

```
@@ -26,9 +26,410 @@
+    # this can generate `/{id}/` -> `//` and `-{id}-` -> `--`, so fix those
+    file_path = file_path.replace("//", "/").replace("--", "_")
+
+    # specialty conversions for existing non-standard usage
+    fixes = {  # old : new
+        ".bam.bai": ".bam.index",
+        "pyclone.tsv": "clonality_pyclone",
+        "copynumber/": "copynumber/copynumber_",
+        "tn_corealigned.bam": "tn_corealigned",
+        "optitype/result": "optitype/optitype_result",
+        "xhla": "optitype/xhla",
+        "/align": "/alignment/align",
+        "sample_summar": "summar",
+        "all_epitopes": "epitopes",
+        ".txt.tn.tsv": ".tsv",
+    }
+    for old, new in fixes.items():
+        file_path = file_path.replace(old, new)
+
+    # special handling for `all_summaries`
+    if key not in ["tumor cimac id", "normal cimac id"]:
+        file_path = file_path.replace("analysis/metrics", "tumor/metrics")
+
+    # split path into pieces
+    file_path = file_path.split("/")
+
+    # specialty conversions for file names / extensions only
+    if "tnscope" in file_path[-1]:
+        temp = file_path[-1].split(".")
+        if temp[-1] != "gz":
+            file_path[-1] = ".".join(temp[-1:] + temp[:-1])
+        else:
+            file_path[-1] = ".".join(temp[-2:] + temp[:-2])
+    if "mosdepth" in file_path[-1]:
+        temp = file_path[-1].split(".")
+        file_path[-1] = ".".join(temp[1:4]) + "_" + temp[0]
+
+    # remove any extra `analysis` at the front
+    if file_path[0] == "analysis":
+        file_path = file_path[1:]
+
+    # look into first step
+    merge_pointer = "0"
+    curr_step, file_path, curr_context = _first_in_context(file_path, context)
+    merge_pointer += "/" + curr_step
+    # then off to the races
+    while len(curr_step):
+        curr_step, file_path, curr_context = _first_in_context(file_path, curr_context)
+        if curr_step:
+            merge_pointer += "/" + curr_step
+
+    return merge_pointer
```

If no paths in `curr_context` match `file_path`, we could end up with `merge_pointer == "0/"` here. Maybe we should raise an error in this case, since we want to find a match for every `file_path` (right?).

---

I think explicitly erroring is better than silently swallowing a non-mergeable entry.

---

is it covered with tests to check that it throughs error for "bad" merge_pointer?